docs/DeadmanAlert.md

```diff
@@ -0,0 +1,99 @@
+### Deadman Alert
+
+The `deadman` function is used in the context of a task to alert when a group stops reporting data.
+For example, say you have the following task definition:
+
+```
+option task = {
+    ...
+    every: 15m,
+    ...
+}
+from(bucket: "telegraf/autogen")
+    |> range(start: -1h)
+    |> filter(fn: (r) => r._measurement == "cpu")
+```
+
+In order to be alerted when the above task stops reporting data for a group, you would define a deadman alert like so:
+```
+option task = {
+    ...
+    every: 15m,
+    ...
+}
+f = (start, stop) =>
+    from(bucket: "telegraf/autogen")
+        |> range(start: start, stop: stop)
+        |> filter(fn: (r) => r._measurement == "cpu")
+
+deadman(f: f, period: 1h, offset: task.every)
+```
+
+where the deadman function is defined as follows:
+```
+deadman = (f, period, offset) => {
```

Comment:
I'm not sure I like this deadman function syntax. It seems to require too many parameters and it seems to be a bit too tied to `from() |> range()`.

I'm curious if there's a way to make it so deadman is chainable. It seems that the way deadman works is it takes two different streams and checks to see if one of the group keys does not exist in the second stream. It seems to me this is possible with just a single call to from. Imagine if instead of using `period` and `offset`, we instead did:

```
deadman = (for) => ...

from(bucket: "telegraf/autogen")
    |> range(start: -period)
    |> filter(fn: (r) => r._measurement == "cpu")
    |> deadman(for: offset)
```

The deadman could be configured so that it uses the `_time` column for a table and it alerts if there are no points within the last `offset` period. It seems to me that this would be the same thing as the current implementation, but it wouldn't require a specific function to accommodate `range()`.

This wouldn't require `group_keys()` or `diff()` to exist since these are already done by the execution engine.

Reply:
@jsternberg I definitely like this signature for deadman. But I'm not sure how you would eliminate `group_keys()` and `diff()`. My goal is to have deadman defined in pure Flux. With this new signature, deadman would be defined as follows:

```
deadman(off, tables=<-) => {
    r = tables
        |> group_keys()

    s = tables
        |> filter(fn: (r) => r._time > now() - off)
        |> group_keys()

    return diff(r:r, s:s) |> alert(crit: (r) => true)
}
```

I think this is a much better definition. However we still need `group_keys()` and `diff()`. Unless you mean for it to be implemented in Go, but I would much rather have it implemented in Flux.

Reply:
I think it could be implemented by adding a new feature to `elapsed()` since `elapsed()` is basically what we want anyway. If we added a new `stop:` parameter to elapsed to indicate an optional stop time, we could do something like this:

```
deadman(tables=<-, for) => elapsed(stop: now()) |> alert(crit: (r) => duration(v: r._value) > for)
```

The `stop` parameter would make it so that elapsed emitted a final row that was the elapsed time since the last entry and the time specified.

Alternatively, maybe this would be better?

```
deadman(tables=<-, for) => limit(n: 1) |> map(fn: (r) => ({r with _value: int(v: now()) - r._time})) |> alert(crit: (r) => duration(v: r._value) > for)
```

This is defined in pure flux and it will take the last row in the table, figure out the difference between the current time and the time of the row, and alert when that difference is greater than the `for` time.

Reply:
@jsternberg definitely like your idea, and I think a good intermediate step would be to change the signature to what you are suggesting. Your solution has a major pro of not triggering the source twice.

Once we have that signature and we add the capability to `elapsed` we can change `deadman`'s implementation.